spec.md

```diff
@@ -1131,6 +1131,8 @@ This RPC will be called by the CO to deprovision a volume.
 This operation MUST be idempotent.
 If a volume corresponding to the specified `volume_id` does not exist or the artifacts associated with the volume do not exist anymore, the Plugin MUST reply `0 OK`.
 
+CSI plugins MUST treat volumes independent from their snapshots.  Controller Plugin MUST support deleting a volume without affecting its existing snapshots, which MUST still be fully operational and acceptable as sources for new volumes as well as appear on `ListSnapshot` calls.
```

Comment:
I think this could be too onerous for some plugins to support, especially if they're not maintaining their own state store independent of any internal store provided by the "real" backend.

I'd rather let the plugin decide how to do this:

1. Fail the volume-delete RPC because it's not allowed there are snapshots.
2. Honor the volume-delete RPC and maintain the existing snapshots (because they're still usable).
3. Honor the volume-delete RPC and remove all linked snapshots.

If there's an ongoing snapshot-related operation in progress (e.g. creating a snapshot from said volume, or creating a volume from a snapshot that's linked to the volume-to-be-removed) then we have error codes a plugin can use to indicate a conflict with an ongoing operation (`ABORTED`).

If (1) is the case then there's already a documented error code for delete's that fail because a resource is in-use (`FAILED_PRECONDITION`).

If the plugin is able to execute (3) without errors, then it's up to the CO to periodically reconcile its internal state w/ that of the plugin in order to account for snapshots that may have disappeared.

Reply:
I agree that we should let the SP decide but this is a good suggestion. So how about `/s/MUST/SHOULD/` in this change to make it a recommendation rather then a requirement?

Reply:
I'm updating the patch to change it to `SHOULD` and also make it contemplate @jdef 's first scenario.  This way plugins can either support deleting volumes independently or fail.  We can leave support for cascade deletion of snapshots as a new feature and leave this patch to just clarify the current expected behavior.